apis/duck/v1alpha1/channelable_types.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/apis/duck/unversioned"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Channelable is a skeleton type wrapping Subscribable and Addressable in the manner we expect resource writers
+// defining compatible resources to embed it. We will typically use this type to deserialize
+// Channelable ObjectReferences and access their subscription and address data.  This is not a real resource.
+type Channelable struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the part where the Channelable fulfills the Subscribable contract.
+	Spec ChannelableSpec `json:"spec,omitempty"`
+
+	Status ChannelableStatus `json:"status,omitempty"`
+}
+
+// ChannelableSpec contains Spec of the Channelable object
+type ChannelableSpec struct {
+	SubscribableTypeSpec `json:",inline"`
+}
+
+// ChannelableStatus contains the Status of a Channelable object.
+type ChannelableStatus struct {
+	// AddressStatus is the part where the Channelable fulfills the Addressable contract.
```

Comment:
FYI, since you're upstreaming, just curious about your thoughts on this?
https://github.com/knative/eventing/issues/1375

I was going to take a crack at this on Monday, but if you're doing it here, consider adding the duckv1beta1.Status field here as well? Currently it's not easy to figure out what the Status of Channelable is without knowing what type it is to get to the actual Status.

Reply:
I think I made a mess of this PR and it needs to be scoped down...

Reply:
I will put a pin in this one and then look at moving the types up one at a time.

`apis/duck/addressable_resolver.go` was the thing that make it hard.